chemprop/v2/data/utils.py

```diff
@@ -2,65 +2,253 @@
+    key_molecule_index: int = 0,
+    seed: int = 0,
+    num_folds: int = 1,
+    args: TrainArgs = None,
+    logger: Logger = None,
+) -> Tuple[MoleculeDataset, MoleculeDataset, MoleculeDataset]:
+    r"""
+    Splits data into training, validation, and test splits.
+
+    :param data: A :class:`~chemprop.data.MoleculeDataset`.
+    :param split_type: Split type.
+    :param sizes: A length-3 tuple with the proportions of data in the train, validation, and test sets.
+    :param key_molecule_index: For data with multiple molecules, this sets which molecule will be considered during splitting.
+    :param seed: The random seed to use before shuffling data.
+    :param num_folds: Number of folds to create (only needed for "cv" split type).
+    :param args: A :class:`~chemprop.args.TrainArgs` object.
+    :param logger: A logger for recording output.
+    :return: A tuple of :class:`~chemprop.data.MoleculeDataset`\ s containing the train,
+             validation, and test splits of the data.
+    """
     if not (len(sizes) == 3 and np.isclose(sum(sizes), 1)):
         raise ValueError(f"Invalid train/val/test splits! got: {sizes}")
 
-    n_train, n_val, n_test = [int(size * len(data)) for size in sizes]
-
-    if split == "random":
-        idxs = np.arange(len(data))
-        np.random.shuffle(idxs)
-
-        train = [data[i] for i in idxs[:n_train]]
-        val = [data[i] for i in idxs[n_train : n_train + n_val]]
-        test = [data[i] for i in idxs[n_train + n_val :]]
-    elif split == "scaffold":
-        scaffold2idxs = defaultdict(set)
-        for i, d in enumerate(data):
-            scaffold = MurckoScaffold.MurckoScaffoldSmiles(mol=d.mol)
-            scaffold2idxs[scaffold].add(i)
-
-        big_index_sets = []
-        small_index_sets = []
-        for idxs in scaffold2idxs.values():
-            if len(idxs) > n_val / 2 or len(idxs) > n_test / 2:
-                big_index_sets.append(idxs)
+    random = Random(seed)
+
+    if args is not None:
+        folds_file, val_fold_index, test_fold_index = (
+            args.folds_file,
+            args.val_fold_index,
+            args.test_fold_index,
+        )
+    else:
+        folds_file = val_fold_index = test_fold_index = None
+
+    # typically include a validation set
+    include_val = True
+    split_fun = train_val_test_split
+    mol_split_fun = train_val_test_split_molecules
+    # default sampling arguments for astartes sampler
+    astartes_kwargs = dict(
+        train_size=sizes[0], test_size=sizes[2], return_indices=True, random_state=seed
+    )
+    # if no validation set, reassign the splitting functions
+    if not sizes[1]:
+        include_val = False
+        split_fun = train_test_split
+        mol_split_fun = train_test_split_molecules
+    else:
+        astartes_kwargs["val_size"] = sizes[1]
+
+    train, val, test = None, None, None
+    if split == "crossval":
+        index_set = args.crossval_index_sets[args.seed]
```

Comment:
What is the difference between using `args.seed` and the input param `seed`?

Reply:
Also if you are allowing `args` to be `None`, these codes can fail if `args` is `None`.

Reply:
args has also now been removed